src/libcore/num/mod.rs

```diff
@@ -2523,42 +2524,120 @@ macro_rules! same_sign_try_from_int_impl {
-                let max = <$signed as FromStrRadixHelper>::max_value() as u128;
-                if u as u128 > max {
-                    Err(TryFromIntError(()))
-                } else {
-                    Ok(u as $signed)
-                }
+            #[inline]
+            fn try_from(u: $source) -> Result<$target, TryFromIntError> {
+                Ok(u as $target)
             }
         }
+    )*}
+}
+
+#[cfg(not(any(
+    target_pointer_width = "16", target_pointer_width = "32", target_pointer_width = "64")))]
+compile_error!("The current implementations of try_from on usize/isize assumes that \
+                the pointer width is either 16, 32, or 64");
+// (source, $(target))
+trivial_try_from_impl!(u8, u8, u16, i16, u32, i32, u64, i64, u128, i128, usize);
+trivial_try_from_impl!(u16, u16, u32, i32, u64, i64, u128, i128, usize);
+trivial_try_from_impl!(u32, u32, u64, i64, u128, i128);
+trivial_try_from_impl!(u64, u64, u128, i128);
+trivial_try_from_impl!(u128, u128);
+trivial_try_from_impl!(usize, usize, u128, i128);
+trivial_try_from_impl!(usize, u64);
+
+trivial_try_from_impl!(i8, i8, i16, i32, i64, i128, isize);
+trivial_try_from_impl!(i16, i16, i32, i64, i128, isize);
+trivial_try_from_impl!(i32, i32, i64, i128);
+trivial_try_from_impl!(i64, i64, i128);
+trivial_try_from_impl!(i128, i128);
+trivial_try_from_impl!(isize, isize, i128);
+trivial_try_from_impl!(isize, i64);
+
+// (target, $(source))
+same_sign_try_from_wider_impl!(u8, u16, u32, u64, u128, usize);
+same_sign_try_from_wider_impl!(i8, i16, i32, i64, i128, isize);
+same_sign_try_from_wider_impl!(u16, u32, u64, u128);
+same_sign_try_from_wider_impl!(i16, i32, i64, i128);
+same_sign_try_from_wider_impl!(u32, u64, u128);
+same_sign_try_from_wider_impl!(i32, i64, i128);
+same_sign_try_from_wider_impl!(u64, u128);
+same_sign_try_from_wider_impl!(i64, i128);
+same_sign_try_from_wider_impl!(usize, u128);
+same_sign_try_from_wider_impl!(isize, u128);
+
+macro_rules! cfg_block {
+    ($(#[$attr:meta]{$($it:item)*})*) => {$($(
+        #[$attr]
+        $it
+    )*)*}
+}
+
+
+cfg_block!(
+    // Platform specific impls for conversions with the same sign:
+    // xsize -> x32
+    // xsize -> x16
+    // x32 -> xsize
+    // x64 -> xsize
+
+    // 16-bit.
+    #[cfg(target_pointer_width = "16")] {
+        // x32 -> xsize
+        same_sign_try_from_wider_impl!(usize, u32);
+        same_sign_try_from_wider_impl!(isize, i32);
+        // xsize -> x16
+        trivial_try_from_impl!(usize, u16);
+        trivial_try_from_impl!(isize, i16);
+    }
+
+    // Same for 16 and 32-bit platforms.
+    #[cfg(any(target_pointer_width = "16", target_pointer_width = "32"))] {
+        // xsize -> x32
+        trivial_try_from_impl!(usize, u32);
+        trivial_try_from_impl!(isize, i32);
+        // x64 -> xsize
+        same_sign_try_from_wider_impl!(usize, u64);
+        same_sign_try_from_wider_impl!(isize, u64);
```

Comment:
Also, I believe the argument order is swapped for `same_sign_try_from_wider_impl` here and in the block above?